File: .skills/openclaw-skills/skills/puckguo/dawang/scripts/feishu_send_image.py
#!/usr/bin/env python3
"""
飞书图片发送工具
用于将本地图片上传到飞书并获取 image_key，然后发送消息
"""

import requests
import sys
import os

def get_tenant_access_token():
    """获取飞书 tenant access token"""
    import json
    
    # 从环境变量获取
    app_id = os.environ.get('FEISHU_APP_ID')
    app_secret = os.environ.get('FEISHU_APP_SECRET')
    
    if not app_id or not app_secret:
        # 尝试从配置文件读取
        config_paths = [
            os.path.expanduser('~/.openclaw/config.json'),
            os.path.expanduser('~/.openclaw/feishu.json'),
        ]
        for config_path in config_paths:
            if os.path.exists(config_path):
                try:
                    with open(config_path, 'r') as f:
                        config = json.load(f)
                        app_id = config.get('feishu', {}).get('appId') or config.get('appId')
                        app_secret = config.get('feishu', {}).get('appSecret') or config.get('appSecret')
                        if app_id and app_secret:
                            break
                except:
                    continue
    
    if not app_id or not app_secret:
        raise ValueError("无法获取 Feishu App ID 和 App Secret，请设置环境变量或配置文件")
    
    url = "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal"
    headers = {"Content-Type": "application/json"}
    data = {"app_id": app_id, "app_secret": app_secret}
    
    response = requests.post(url, headers=headers, json=data)
    result = response.json()
    
    if result.get("code") != 0:
        raise Exception(f"获取 token 失败: {result}")
    
    return result["tenant_access_token"]

def upload_image(image_path, token):
    """上传图片到飞书，获取 image_key"""
    url = "https://open.feishu.cn/open-apis/im/v1/images"
    
    headers = {
        "Authorization": f"Bearer {token}"
    }
    
    with open(image_path, 'rb') as f:
        files = {
            'image': (os.path.basename(image_path), f, 'image/png')
        }
        data = {
            'image_type': 'message'
        }
        
        response = requests.post(url, headers=headers, files=files, data=data)
    
    result = response.json()
    
    if result.get("code") != 0:
        raise Exception(f"上传图片失败: {result}")
    
    return result["data"]["image_key"]

def send_image_message(image_key, chat_id, token):
    """发送图片消息"""
    url = "https://open.feishu.cn/open-apis/im/v1/messages"
    
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }
    
    params = {
        "receive_id_type": "chat_id"
    }
    
    # 构建消息内容
    content = {
        "image_key": image_key
    }
    
    import json
    data = {
        "receive_id": chat_id,
        "msg_type": "image",
        "content": json.dumps(content)
    }
    
    response = requests.post(url, headers=headers, params=params, json=data)
    result = response.json()
    
    if result.get("code") != 0:
        raise Exception(f"发送消息失败: {result}")
    
    return result

def main():
    if len(sys.argv) < 3:
        print("用法: python3 feishu_send_image.py <图片路径> <聊天ID>")
        print("示例: python3 feishu_send_image.py /path/to/image.png ou_xxx")
        sys.exit(1)
    
    image_path = sys.argv[1]
    chat_id = sys.argv[2]
    
    # 检查文件是否存在
    if not os.path.exists(image_path):
        print(f"错误: 文件不存在 {image_path}")
        sys.exit(1)
    
    try:
        # 获取 token
        print("正在获取 access token...")
        token = get_tenant_access_token()
        print("✓ 获取 token 成功")
        
        # 上传图片
        print(f"正在上传图片: {image_path}")
        image_key = upload_image(image_path, token)
        print(f"✓ 图片上传成功，image_key: {image_key}")
        
        # 发送消息
        print(f"正在发送消息到: {chat_id}")
        result = send_image_message(image_key, chat_id, token)
        print(f"✓ 消息发送成功!")
        print(f"  Message ID: {result['data']['message_id']}")
        
        # 输出 image_key 供其他工具使用
        print(f"\nIMAGE_KEY:{image_key}")
        
    except Exception as e:
        print(f"错误: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/kagura-agent/agent-memes/scripts/feishu-send-image.mjs
#!/usr/bin/env node
// Usage: node scripts/feishu-send-image.mjs <target> <image_path>
// target: "user:open_id_xxx" or "oc_xxx" (chat_id)

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve } from 'path';
import { homedir } from 'os';

const API = 'https://open.feishu.cn/open-apis';
// Secure token cache: user-private directory instead of world-readable /tmp
const CACHE_DIR = resolve(homedir(), '.cache/agent-memes');
const TOKEN_CACHE = resolve(CACHE_DIR, 'feishu-token.json');

const [target, imagePath] = process.argv.slice(2);
if (!target || !imagePath) {
  console.error('Usage: node scripts/feishu-send-image.mjs <target> <image_path>');
  process.exit(1);
}

// --- credentials via centralized helper ---
import { execSync } from 'child_process';
import { mkdirSync } from 'fs';

let appId = process.env.FEISHU_APP_ID;
let appSecret = process.env.FEISHU_APP_SECRET;

if (!appId || !appSecret) {
  try {
    const helperPath = resolve(import.meta.dirname || new URL('.', import.meta.url).pathname, 'get-credential.sh');
    const creds = execSync(`bash "${helperPath}" feishu`, { encoding: 'utf8' }).trim();
    const parts = creds.split(' ');
    appId = parts[0];
    appSecret = parts[1];
  } catch {}
}
if (!appId || !appSecret) {
  console.error('Missing credentials. Set FEISHU_APP_ID + FEISHU_APP_SECRET env vars, or configure ~/.openclaw/openclaw.json');
  process.exit(1);
}

// Ensure secure cache directory exists (0700)
try { mkdirSync(CACHE_DIR, { recursive: true, mode: 0o700 }); } catch {}

// --- tenant_access_token (cached) ---
async function getToken() {
  if (existsSync(TOKEN_CACHE)) {
    try {
      const cached = JSON.parse(readFileSync(TOKEN_CACHE, 'utf8'));
      if (cached.token && cached.expiresAt > Date.now() + 60_000) return cached.token;
    } catch {}
  }
  const res = await fetch(`${API}/auth/v3/tenant_access_token/internal`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ app_id: appId, app_secret: appSecret }),
  });
  const data = await res.json();
  if (data.code !== 0) throw new Error(`token error: ${data.msg}`);
  const token = data.tenant_access_token;
  writeFileSync(TOKEN_CACHE, JSON.stringify({ token, expiresAt: Date.now() + data.expire * 1000 }));
  return token;
}

// --- upload image ---
async function uploadImage(token, filePath) {
  const absPath = resolve(filePath);
  const bytes = readFileSync(absPath);
  const blob = new Blob([bytes]);
  const form = new FormData();
  form.append('image_type', 'message');
  form.append('image', blob, absPath.split('/').pop());
  const res = await fetch(`${API}/im/v1/images`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: form,
  });
  const data = await res.json();
  if (data.code !== 0) throw new Error(`upload error: ${data.code} ${data.msg}`);
  return data.data.image_key;
}

// --- send message ---
async function sendImage(token, target, imageKey) {
  let receiveIdType, receiveId;
  if (target.startsWith('user:')) {
    receiveIdType = 'open_id';
    receiveId = target.slice(5);
  } else {
    receiveIdType = 'chat_id';
    receiveId = target;
  }
  const res = await fetch(`${API}/im/v1/messages?receive_id_type=${receiveIdType}`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ receive_id: receiveId, msg_type: 'image', content: JSON.stringify({ image_key: imageKey }) }),
  });
  const data = await res.json();
  if (data.code !== 0) throw new Error(`send error: ${data.code} ${data.msg}`);
  return data.data.message_id;
}

// --- main ---
try {
  const token = await getToken();
  console.log('token ok');
  const imageKey = await uploadImage(token, imagePath);
  console.log('uploaded:', imageKey);
  const msgId = await sendImage(token, target, imageKey);
  console.log('sent:', msgId);
} catch (e) {
  console.error(e.message);
  process.exit(1);
}
